### PR TITLE
Fixed #26351 -- Added database check about MySQL sql_mode variable

### PR DESCRIPTION
**Pull request by claudep**
Thanks Adam Chainz for the initial implementation in django-mysql.

### CODE REVIEW

**Comment by timgraham** [comment]
The code should likely live in the database backend (`validation.py` contains database-specific field checks, for example). If possible, I'd like to see something that avoids the
```
for conn in connections.all():
    if not hasattr(conn, 'mysql_version'):
        continue
```
boilerplate in all checks of this type.

 Documentation to be updated is in `docs/ref/checks.txt`.

**Comment by adamchainz** [comment]
:smile:

**Comment by adamchainz** [comment]
Btw just linking to the MySQL docs isn't necessarily the most helpful for an immediate fix, it naturally leads one to google "django set sql_mode" etc. , why not add a docs page on how to set it in the connection `OPTIONS` [like django-mysql has](https://django-mysql.readthedocs.org/en/latest/checks.html#django-mysql-w001-strict-mode) ?

**Comment by claudep** [comment]
Yes, that's a good idea. However, as stated on the ticket, the main question now is to decide how we will apply database-related checks.

File: docs/ref/checks.txt [hunk 1]
```diff
@@ -84,6 +84,14 @@ Django's system checks are organized using the following tags:
 * ``templates``: Checks template related configuration.
 * ``caches``: Checks cache related configuration.
 * ``urls``: Checks URL configuration.
+* ``database``: Checks any database-related connection issues. Database checks
+  are not run by default. They are only explicitly run by the
+  :djadmin:`migrate` command or if you specify the ``database`` tag when calling
+  the :djadmin:`migrate` command.
```
Comment (claudep) [inline docs/ref/checks.txt]: Sorry, this should be the `check` command, of course.

Comment (shaib) [inline docs/ref/checks.txt]: I think you meant  the test command

File: django/core/management/base.py [hunk 2]
```diff
@@ -360,6 +360,9 @@ def execute(self, *args, **options):
                 translation.activate(saved_locale)
         return output
 
+    def _check_issues(self, **kwargs):
```
Comment (timgraham) [inline django/core/management/base.py]: I think I'd find a name like `run_checks()` more intuitive (or with a leading underscore if you like).

File: django/db/backends/mysql/validation.py [hunk 3]
```diff
@@ -3,6 +3,23 @@
 
 
 class DatabaseValidation(BaseDatabaseValidation):
+    def check(self, **kwargs):
+        warnings = super(DatabaseValidation, self).check(**kwargs)
```
Comment (timgraham) [inline django/db/backends/mysql/validation.py]: "warnings" -> "issues" for consistency with elsewhere?
I would also create a separate "check_sql_mode()" method so if more checks are added, we don't end up with a giant method (similar to the pattern in `Field.check()`).

File: django/db/backends/mysql/validation.py [hunk 4]
```diff
@@ -3,6 +3,23 @@
...
+            cursor.execute("""SELECT @@sql_mode""")
+            sql_mode = cursor.fetchone()
+        modes = set(sql_mode[0].split(','))
+        if not (modes & {'STRICT_TRANS_TABLES', 'STRICT_ALL_TABLES', 'TRADITIONAL'}):
+            warnings.append(checks.Warning(
+                "MySQL Strict Mode is not set for database connection '%s'" % self.connection.alias,
+                hint="MySQL's Strict Mode fixes many data integrity problems in MySQL, "
+                     "such as data truncation upon insertion, by escalating warnings into "
+                     "errors. It is strongly recommended you activate it. See: "
+                     "https://docs.djangoproject.com/en/stable/ref/databases/#mysql-sql-mode",
```
Comment (timgraham) [inline django/db/backends/mysql/validation.py]: stable -> get_docs_version() ?

File: docs/ref/databases.txt [hunk 5]
```diff
@@ -439,6 +439,24 @@ more details.
...
 
+.. _mysql-sql-mode:
+
+Setting ``sql_mode``
+~~~~~~~~~~~~~~~~~~~~
+
+From MySQL 5.7 onwards, the default value of the ``sql_mode`` option contains
+``STRICT_TRANS_TABLES``. That option is escalating warnings into errors when
+data are truncated upon insertion, so Django highly recommends activating a
+`strict mode`_ for MySQL to prevent data losses (either ``STRICT_TRANS_TABLES``,
```
Comment (timgraham) [inline docs/ref/databases.txt]: data loss

File: docs/ref/databases.txt [hunk 6]
```diff
@@ -439,6 +439,24 @@ more details.
 .. _MySQL option file: https://dev.mysql.com/doc/refman/5.6/en/option-files.html
 .. _MySQLdb documentation: http://mysql-python.sourceforge.net/
 
+.. _mysql-sql-mode:
+
+Setting ``sql_mode``
+~~~~~~~~~~~~~~~~~~~~
+
+From MySQL 5.7 onwards, the default value of the ``sql_mode`` option contains
+``STRICT_TRANS_TABLES``. That option is escalating warnings into errors when
```
Comment (timgraham) [inline docs/ref/databases.txt]: That option escalates warnings

File: docs/ref/databases.txt [hunk 7]
```diff
@@ -439,6 +439,24 @@ more details.
 .. _MySQL option file: https://dev.mysql.com/doc/refman/5.6/en/option-files.html
 .. _MySQLdb documentation: http://mysql-python.sourceforge.net/
 
+.. _mysql-sql-mode:
+
+Setting ``sql_mode``
+~~~~~~~~~~~~~~~~~~~~
+
+From MySQL 5.7 onwards, the default value of the ``sql_mode`` option contains
```
Comment (adamchainz) [inline docs/ref/databases.txt]: N.B. It's also the default in fresh installs of MySQL 5.6 (not on top of existing data from 5.5). See morgo's comment on https://code.djangoproject.com/ticket/15940

File: docs/ref/databases.txt [hunk 8]
```diff
@@ -439,6 +439,24 @@ more details.
...
+``STRICT_TRANS_TABLES``. That option is escalating warnings into errors when
+data are truncated upon insertion, so Django highly recommends activating a
+`strict mode`_ for MySQL to prevent data losses (either ``STRICT_TRANS_TABLES``,
+``STRICT_ALL_TABLES``, or ``TRADITIONAL``).
+
+.. _strict mode: https://dev.mysql.com/doc/refman/5.7/en/sql-mode.html#sql-mode-strict
+
+If you are using an older version of MySQL, you can set the ``sql_mode`` like
+other MySQL options, that is either in a config file, or with the entry
```
Comment (timgraham) [inline docs/ref/databases.txt]: options: either in a config file or with (no comma)

File: tests/check_framework/test_database.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+            checks.run_checks(tags=["database"])
+            self.assertTrue(mocked_check.called)
+
+    @unittest.skipUnless(connection.vendor == 'mysql', "Test only for MySQL")
+    def test_mysql_strict_mode(self):
+        good_sql_modes = [
+            'STRICT_TRANS_TABLES,STRICT_ALL_TABLES',
+            'STRICT_TRANS_TABLES',
+            'STRICT_ALL_TABLES',
+            'TRADITIONAL',
```
Comment (adamchainz) [inline tests/check_framework/test_database.py]: There's no need to accept "Traditional". When you *set* traditional, MySQL sets multiple modes, including `STRICT_TRANS_TABLES`. If you then ask it back what `sql_mode` is, you get them all back:

```
adamj@localhost [1]> set sql_mode = 'traditional';
Query OK, 0 rows affected (0.01 sec)

adamj@localhost [2]> select @@sql_mode;
+------------------------------------------------------------------------------------------------------------------------------------------------------+
| @@sql_mode                                                                                                                                           |
+------------------------------------------------------------------------------------------------------------------------------------------------------+
| STRICT_TRANS_TABLES,STRICT_ALL_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,TRADITIONAL,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION |
+------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

File: docs/ref/databases.txt [hunk 10]
```diff
@@ -439,6 +439,24 @@ more details.
...
+.. _mysql-sql-mode:
+
+Setting ``sql_mode``
+~~~~~~~~~~~~~~~~~~~~
+
+From MySQL 5.7 onwards, the default value of the ``sql_mode`` option contains
+``STRICT_TRANS_TABLES``. That option is escalating warnings into errors when
+data are truncated upon insertion, so Django highly recommends activating a
+`strict mode`_ for MySQL to prevent data losses (either ``STRICT_TRANS_TABLES``,
+``STRICT_ALL_TABLES``, or ``TRADITIONAL``).
```
Comment (adamchainz) [inline docs/ref/databases.txt]: as below, I wouldn't mention TRADITIONAL, as it just maps on to a selection of modes, including both `STRICT_TRANS_TABLES` and `STRICT_ALL_TABLES`.

File: django/db/backends/mysql/validation.py [hunk 11]
```diff
@@ -3,6 +3,23 @@
 
 
 class DatabaseValidation(BaseDatabaseValidation):
+    def check(self, **kwargs):
+        warnings = super(DatabaseValidation, self).check(**kwargs)
+        with self.connection.cursor() as cursor:
+            cursor.execute("""SELECT @@sql_mode""")
+            sql_mode = cursor.fetchone()
+        modes = set(sql_mode[0].split(','))
+        if not (modes & {'STRICT_TRANS_TABLES', 'STRICT_ALL_TABLES', 'TRADITIONAL'}):
```
Comment (adamchainz) [inline django/db/backends/mysql/validation.py]: as below, no need to check for `TRADITIONAL`

File: docs/ref/databases.txt [hunk 12]
```diff
@@ -439,6 +439,24 @@ more details.
...
+
+.. _strict mode: https://dev.mysql.com/doc/refman/5.7/en/sql-mode.html#sql-mode-strict
+
+If you are using an older version of MySQL, you can set the ``sql_mode`` like
```
Comment (timgraham) [inline docs/ref/databases.txt]: I understand the intent but I guess "If you are using an older version of MySQL" isn't really needed since the instructions work on any version (maybe you want to say something like "If you need to customize the SQL mode, you can set it like...").

File: docs/ref/databases.txt [hunk 13]
```diff
@@ -439,6 +439,24 @@ more details.
 .. _MySQL option file: https://dev.mysql.com/doc/refman/5.6/en/option-files.html
 .. _MySQLdb documentation: http://mysql-python.sourceforge.net/
 
+.. _mysql-sql-mode:
```
Comment (timgraham) [inline docs/ref/databases.txt]: Could you make this note a separate commit (perhaps resolving https://code.djangoproject.com/ticket/15940?) so we can backport it?

File: docs/ref/checks.txt [hunk 14]
```diff
@@ -84,6 +84,14 @@ Django's system checks are organized using the following tags:
 * ``templates``: Checks template related configuration.
 * ``caches``: Checks cache related configuration.
 * ``urls``: Checks URL configuration.
+* ``database``: Checks any database-related connection issues. Database checks
+  are not run by default. They are only explicitly run by the
```
Comment (timgraham) [inline docs/ref/checks.txt]: chop "explicitly"

File: docs/ref/checks.txt [hunk 15]
```diff
@@ -574,6 +582,8 @@ If you're using MySQL, the following checks will be performed:
 
 * **mysql.E001**: MySQL does not allow unique ``CharField``\s to have a
   ``max_length`` > 255.
+* **mysql.W002**: MySQL Strict Mode is not set for database connection
```
Comment (timgraham) [inline docs/ref/checks.txt]: I think the numbering would start at W001 -- did you see any contradictions?

File: django/db/backends/mysql/validation.py [hunk 16]
```diff
@@ -3,6 +3,23 @@
 
 
 class DatabaseValidation(BaseDatabaseValidation):
+    def check(self, **kwargs):
+        warnings = super(DatabaseValidation, self).check(**kwargs)
+        with self.connection.cursor() as cursor:
+            cursor.execute("""SELECT @@sql_mode""")
```
Comment (timgraham) [inline django/db/backends/mysql/validation.py]: are triple quotes needed?

File: tests/check_framework/test_database.py [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+        """
+        def f1(**kwargs):
+            return [5]
+
+        registry = CheckRegistry()
+        registry.register("database")(f1)
+        errors = registry.run_checks()
+        self.assertEqual(errors, [])
+
+        errors2 = registry.run_checks(tags=["database"])
```
Comment (timgraham) [inline tests/check_framework/test_database.py]: use single quotes throughout?

File: docs/ref/checks.txt (hunk 15 above)
Comment (claudep) [inline docs/ref/checks.txt]: I checked https://docs.djangoproject.com/en/1.9/ref/checks/ and saw that error numbers are not reused for warnings, and conversely.

**Comment by claudep** [comment]
buildbot, test on precise.

**Comment by claudep** [comment]
Failures on precise are unrelated.

File: docs/ref/checks.txt [hunk 18]
```diff
@@ -84,6 +84,14 @@ Django's system checks are organized using the following tags:
 * ``templates``: Checks template related configuration.
 * ``caches``: Checks cache related configuration.
 * ``urls``: Checks URL configuration.
+* ``database``: Checks any database-related connection issues. Database checks
```
Comment (shaib) [inline docs/ref/checks.txt]: This is not quite in agreement with the code. Connection issues are not checked AFAICT, and in fact nothing is checked except on MySql. Maybe "Checks database-related configuration issues"?

File: docs/topics/checks.txt [hunk 19]
```diff
@@ -120,17 +120,22 @@ The code below is equivalent to the code above::
...
 
 In some cases, you won't need to register your check function -- you can
 piggyback on an existing registration.
 
-Fields, models, and model managers all implement a ``check()`` method that is
-already registered with the check framework. If you want to add extra checks,
-you can extend the implementation on the base class, perform any extra
-checks you need, and append any messages to those generated by the base class.
-It's recommended that you delegate each check to separate methods.
+Fields, models, model managers, and database connections all implement a
```
Comment (shaib) [inline docs/topics/checks.txt]: database backends

File: docs/topics/checks.txt [hunk 20]
```diff
@@ -120,17 +120,22 @@ The code below is equivalent to the code above::
...
+``check()`` method that is already registered with the check framework. If you
+want to add extra checks, you can extend the implementation on the base class,
+perform any extra checks you need, and append any messages to those generated
+by the base class. It's recommended that you delegate each check to separate
+methods.
+
+.. versionchanged:: 1.10
+
+    Database connection checks were added.
```
Comment (shaib) [inline docs/topics/checks.txt]: ditto

**Comment by shaib** [comment]
Other than minor doc notes, LGTM

File: docs/ref/checks.txt (hunk 18 above)
Comment (timgraham) [inline docs/ref/checks.txt]: Shai's wording sounds good to me. Should we rename the similarly name function `check_database_connections` as well?

File: django/core/checks/registry.py [hunk 21]
```diff
@@ -70,6 +71,11 @@ def run_checks(self, app_configs=None, tags=None, include_deployment_checks=Fals
         if tags is not None:
             checks = [check for check in checks
                       if hasattr(check, 'tags') and set(check.tags) & set(tags)]
+        else:
+            # By default, 'database'-tagged checks are not run as they do more
+            # than mere static code analysis.
+            checks = [check for check in checks
+                      if not hasattr(check, 'tags') or 'database' not in check.tags]
```
Comment (timgraham) [inline django/core/checks/registry.py]: 'database' -> `Tags.database`

File: django/core/management/commands/migrate.py [hunk 22]
```diff
@@ -45,6 +46,11 @@ def add_arguments(self, parser):
         parser.add_argument('--run-syncdb', action='store_true', dest='run_syncdb',
             help='Creates tables for apps without migrations.')
 
+    def _run_checks(self, **kwargs):
+        issues = checks.run_checks(tags=['database'])
```
Comment (timgraham) [inline django/core/management/commands/migrate.py]: Tags.database

File: django/db/backends/mysql/validation.py [hunk 23]
```diff
@@ -1,8 +1,31 @@
...
+            sql_mode = cursor.fetchone()
+        modes = set(sql_mode[0].split(','))
+        if not (modes & {'STRICT_TRANS_TABLES', 'STRICT_ALL_TABLES'}):
+            return [checks.Warning(
+                "MySQL Strict Mode is not set for database connection '%s'" % self.connection.alias,
+                hint="MySQL's Strict Mode fixes many data integrity problems in MySQL, "
+                     "such as data truncation upon insertion, by escalating warnings into "
+                     "errors. It is strongly recommended you activate it. See: "
+                     "https://docs.djangoproject.com/en/%s/ref/databases/#mysql-sql-mode" % (
+                         get_docs_version(),),
```
Comment (timgraham) [inline django/db/backends/mysql/validation.py]: Would look better with the `% (` on the next line I think. `% (get_docs_version(),),`

File: docs/ref/checks.txt [hunk 24]
```diff
@@ -84,6 +84,14 @@ Django's system checks are organized using the following tags:
...
+  are not run by default. They are only run by the :djadmin:`migrate` command
```
Comment (timgraham) [inline docs/ref/checks.txt]: We could add the rationale "are not run by default because they do more than static code analysis as regular checks do."

File: tests/check_framework/test_database.py [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+
+    def test_database_checks_not_run_by_default(self):
+        """
+        `database` checks are only run when their tag is specified.
+        """
+        def f1(**kwargs):
+            return [5]
+
+        registry = CheckRegistry()
+        registry.register('database')(f1)
```
Comment (timgraham) [inline tests/check_framework/test_database.py]: (replace all 'database' with Tags.database)

File: tests/check_framework/test_database.py [hunk 26]
```diff
@@ -0,0 +1,57 @@
...
+            'STRICT_TRANS_TABLES,STRICT_ALL_TABLES',
+            'STRICT_TRANS_TABLES',
+            'STRICT_ALL_TABLES',
+        ]
+        for response in good_sql_modes:
+            with mock.patch('django.db.backends.utils.CursorWrapper.fetchone', create=True,
+                    return_value=(response,)):
+                self.assertEqual(self.func(None), [])
+
+        bad_sql_modes = [
```
Comment (timgraham) [inline tests/check_framework/test_database.py]: could single line this

**Comment by adamchainz** [comment]
🎺
